### ISSUE TITLE
Inconsistent behavior of handleRedirectPromise causes spurious errors

### ISSUE THREAD

**Issue by patrick-jones** [AzureAD/microsoft-authentication-library-for-js#2403]
>> Please follow the issue template below. Failure to do so will result in a delay in answering your question.

## Library
- [ ] `msal@1.x.x` or `@azure/msal@1.x.x`
- [x] `@azure/msal-browser@2.1.0`
- [ ] `@azure/msal-angular@0.x.x`
- [ ] `@azure/msal-angular@1.x.x`
- [ ] `@azure/msal-angularjs@1.x.x`

**Important**: Please fill in your exact version number above, e.g. `msal@1.1.3`.

## Description

Depending on the configuration value of `auth.navigateToLoginRequestUrl` and your redirect url, `handleRedirectPromise` may (or may not) return `null` when there is actually valid authentication data present. `handleRedirectPromise` actually redirects the browser to a new location and returns `null`. However, the redirect does not happen fast enough to avoid client code from seeing the `null` return value and initiating a new login redirect. This results in a `BrowserAuthError` with the code `interaction_in_progress`.

The heart of the problem is the idea that async code that calls a redirect does not need to return any result because you're leaving the page anyway. However, in practice the redirect may take long enough that the javascript code continues to execute with invalid results, leading to brief flashes of spurious error conditions.

## Security
- [ ] Is this issue security related?

## Regression
- [ ] Did this behavior work before? 
Version:

Unknown

## MSAL Configuration

```js
// Provide configuration values here.
// For Azure B2C issues, please include your policies.
const msalConfig = {
    auth: {
        clientId: 'XXXXXXXXXX',
        authority: 'https://login.microsoftonline.com/XXXXXXXXXXXXX',
        redirectUri: `${window.location.origin}/learn/oauth`,
        // navigateToLoginRequestUrl is true by default
    },
    cache: {
        cacheLocation: 'sessionStorage',
        storeAuthStateInCookie: false,
    },
};

```

## Reproduction steps
```js
// Provide relevant code snippets here.
// For Azure B2C issues, please include your policies.

// inside authManager.js we have a login() function that returns Promise<AccountInfo
const login = (loginRequest = options.loginRequest, interactionType = loginInteractionType) => {
    return _authApp.handleRedirectPromise() .then(handleResponse(loginRequest, interactionType));
};

// note overall similarity to sample code at
// https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/38bb90e4e94a8d89363d6266d91390d33910d2f0/samples/msal-browser-samples/VanillaJSTestApp2.0/app/onPageLoad/auth.js#L23
const handleResponse = (loginRequest, interactionType) => (resp) => {
    const account = resp ? resp.account : getAccount(); // getAccount is basically getAllAccounts()[0]
    if (account) { // <--- PROBLEM: resp is null, account is null, so we call _authApp.loginRedirect(loginRequest) again
        return account;
    }

    if (interactionType === 'popup') {
        return _authApp.loginPopup(loginRequest).then((authResult) => authResult.account);
    } else {
        return _authApp.loginRedirect(loginRequest);
        // since the previous line also resolves to undefined instead on an AccountInfo, we actually include this workaround
        // to prevent the site from briefly rendering with a null account:
        // return _authApp.loginRedirect(loginRequest).then(() => new Promise((resolve) => setTimeout(resolve, 60000))).then(() => { throw new Error('Redirect failed') })
    }
};

// inside site index.js
// we often briefly see the interaction_in_progress error here
authManager
    .login()
    .then((account) => {
        ReactDOM.render(<App user={account} />, container);
    })
    .catch((error) => {
        ReactDOM.render(
            <div>Authentication failed: {error.toString()}.</div>,
            container
        );
    });
```

The problem is caused by:
* The user is signing in for the first time
* `auth.navigateToLoginRequestUrl` is `true`
* Your redirect page is not the same as the page you initially visited to initiate the redirect (for example, the redirect page is /oauth but the user visited /home)

The problem is here: https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/679dab0146a70dde87dee58781bcbcdadf20397d/lib/msal-browser/src/app/ClientApplication.ts#L136

Given these conditions, control falls to the second `else if` on line 145. At this point there is a valid response hash available, but the code calls `BrowserUtils.navigateWindow(homepage, true);` without returning the processed hash. Again, the assumption is that you're leaving the page anyway so you don't need to return `this.handleHash(responseHash)`. But in practice this leads to spurious errors.

## Expected behavior

No spurious error conditions.

## Browsers/Environment
- [x ] Chrome
- [ ] Firefox
- [ ] Edge
- [ ] Safari
- [ ] IE
- [ ] Other (Please add browser name here)

**Comment by sameerag** [comment]
@tnorling Can you please take a look at this?

**Comment by tnorling** [comment]
@patrick-jones Thank you for the detailed explanation of the problem! It sounds like this would be solved by putting a delay between kicking off navigation and returning from `handleRedirectPromise`. Is that a reasonable conclusion? I'll see if I can reproduce this and get a PR up for a fix.

**Comment by patrick-jones** [comment]
I would think that you could just return `this.handleHash(responseHash)` after calling `BrowserUtils.navigateWindow(homepage, true)`. Maybe it won't complete before the redirect happens, but at least if it does it's the correct return value.

**Comment by tnorling** [comment]
@patrick-jones A fix for this will be included in version 2.4.0 which will be released shortly. Thanks again for bringing this to our attention!
